### PR TITLE
Bug 1288463 – Support tab reordering in the tab tray

### PR DESCRIPTION
**Pull request by varkor**
You can now start reärranging your tabs by long-pressing on a tab, and then moving it around. iOS 9+.

### CODE REVIEW

File: Client/Frontend/Browser/TabTrayController.swift [hunk 1]
```diff
@@ -472,6 +490,48 @@ class TabTrayController: UIViewController {
...
+                }
+                self.collectionView.beginInteractiveMovementForItemAtIndexPath(indexPath)
+                for item in 0..<self.tabDataSource.collectionView(self.collectionView, numberOfItemsInSection: 0) {
+                    guard let cell = self.collectionView.cellForItemAtIndexPath(NSIndexPath(forItem: item, inSection: 0)) as? TabCell else {
+                        continue
+                    }
+                    if item == indexPath.item {
+                        UIView.animateWithDuration(0.2, delay: 0, options: [.AllowUserInteraction, .BeginFromCurrentState], animations: {
+                            cell.transform = CGAffineTransformScale(CGAffineTransformIdentity, 1.05, 1.05)
+                            cell.alpha = 0.9
```
Comment (thebnich) [inline Client/Frontend/Browser/TabTrayController.swift]: Nit: Let's extract all these magic constants to `TabTrayControllerUX`.

File: Client/Frontend/Browser/TabManager.swift [hunk 2]
```diff
@@ -270,6 +270,23 @@ class TabManager : NSObject {
         configureTab(tab, request: request, afterTab: afterTab, flushToDisk: flushToDisk, zombie: zombie)
         return tab
     }
+    
+    func moveTab(isPrivate privateMode: Bool, fromIndex visibleFromIndex: Int, toIndex visibleToIndex: Int) {
+        assert(NSThread.isMainThread())
+        
+        let fromIndex = tabs.indexOf((privateMode ? privateTabs : normalTabs)[visibleFromIndex]) ?? tabs.count - 1
```
Comment (thebnich) [inline Client/Frontend/Browser/TabManager.swift]: Nit: factor out a `let currentTabs = privateMode ? privateTabs : normalTabs`

**Comment by thebnich** [comment]
This feels really nice! I did a find a bug: If you long press a tab and then release without dragging, that tab will stay expanded/partially transparent.

One other (minor) tweak that would be nice might be to remember the view-relative offset of the touch down position, then add that offset to the drag position. That way, if I long press a tab in the corner, the middle of the tab won't snap to my finger when I move.

**Comment by varkor** [comment]
I've implemented the drag position tweak you suggested — you're right, I do think it works a little better this way.

**Pull request merged**